src/pool/srv_pool.c

```diff
@@ -2881,14 +2878,18 @@ ds_pool_connect_handler(crt_rpc_t *rpc, int handler_version)
 					    &out->pco_space);
 out_map_version:
 	out->pco_op.po_map_version = ds_pool_get_version(svc->ps_pool);
-	if (map_buf)
-		D_FREE(map_buf);
-	if (hdl)
-		D_FREE(hdl);
-	D_FREE(machine);
 out_lock:
 	ABT_rwlock_unlock(svc->ps_lock);
 	rdb_tx_end(&tx);
+	if (transfer_map) {
+		rc2 = ds_pool_transfer_map_buf(map_buf, map_version, rpc, in->pci_map_bulk,
+					       &out->pco_map_buf_size);
+		if (rc == 0)
+			rc = rc2;
```

Comment:
if rc from rdb_tx_commit() is nonzero, then I guess we would ideally want to avoid transferring the pool map buffer to the client.

A potential other problem is that when rdb_tx_commit() has been done (successfully, returning rc==0), then it is as if the connect has succeeded. However, if this transfer subsequently fails, the results of the already-committed transaction would need to be rolled back (otherwise the PS thinks it has one more pool handle than the client thinks has been successfully connected). The comment in the original code mentions transferring the pool map before committing updates for that reason. I guess the existing code that performs pool_space_query_bcast() after committing the tx has a similar problem (client will get a nonzero rc in the RPC reply, yet the PS thinks the client is connected).

Reply:
yea i see the issue. unfortunately i do not know what would be required to unroll the tx in case the transfer fails in this case. i was just experimenting to see if moving the transfer to outside the lock works, which seems like it does, but one needs to handle the case when the transfer fails.

Reply:
it sounds like there are other places like ds_pool_prop_set_handler also where a similar issue exists